# **Sprint 3: DataFrames y análisis estadístico**

## **Tarea M3 T02**



In [7]:
#Ejercicio 1:Descarga el fichero Airlines Delay: Airline on-time statistics and delay causas y cárgalo en un Pandas Dataframe.
# Explora los datos que contiene, explica brevemente qué variables existen y queda únicamente con las columnas que consideres relevantes. Justifica tu elección.


#1.1: DataFrame completo:

import pandas as pd 
df = pd.read_csv("C:/Users/acbon/Desktop/Data Science/DelayedFlights.csv")
print(df.head)


<bound method NDFrame.head of          Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
0                 0  2008      1           3          4   2003.0        1955   
1                 1  2008      1           3          4    754.0         735   
2                 2  2008      1           3          4    628.0         620   
3                 4  2008      1           3          4   1829.0        1755   
4                 5  2008      1           3          4   1940.0        1915   
...             ...   ...    ...         ...        ...      ...         ...   
1936753     7009710  2008     12          13          6   1250.0        1220   
1936754     7009717  2008     12          13          6    657.0         600   
1936755     7009718  2008     12          13          6   1007.0         847   
1936756     7009726  2008     12          13          6   1251.0        1240   
1936757     7009727  2008     12          13          6   1110.0        1103   

         

In [8]:
#1.2: DataFrame reducido:

import pandas as pd 
df = pd.read_csv("C:/Users/acbon/Desktop/Data Science/DelayedFlights.csv")
random_seed = 42
df_reducido = df.sample(n=200000, random_state=random_seed)
print(df_reducido.head())
print(df_reducido.columns)

         Unnamed: 0  Year  Month  DayofMonth  DayOfWeek  DepTime  CRSDepTime  \
1782417     6583804  2008     12           6          6   1805.0        1755   
512712      1604365  2008      3          23          7   2120.0        1200   
447137      1395655  2008      3          31          1   1731.0        1710   
55082        178786  2008      1           6          7   1507.0        1413   
877634      2972767  2008      5          22          4    637.0         630   

         ArrTime  CRSArrTime UniqueCarrier  ...  TaxiIn TaxiOut  Cancelled  \
1782417   2146.0        2035            YV  ...     5.0    99.0          0   
512712    2241.0        1333            NW  ...     7.0    19.0          0   
447137    1819.0        1810            OO  ...     3.0     1.0          0   
55082     1601.0        1509            OO  ...     5.0    17.0          0   
877634     812.0         750            CO  ...     5.0    16.0          0   

         CancellationCode  Diverted  CarrierDelay 

Este conjunto de datos contiene información sobre vuelos del año 2008 y proporciona una variedad de estadísticas y datos relacionados con esos vuelos. Algunas de las informaciones relevantes que se pueden extraer de este conjunto de datos incluyen:

-Datos sobre vuelos del año 2008.

-Distribución temporal de vuelos a lo largo del año, meses y días de la semana.

-Promedio de retrasos en despegue y aterrizaje.

-Distancia promedio de vuelo.

-Tiempos promedio de taxi de entrada y salida.

-Frecuencia de cancelaciones y desvíos.

-Causas de retrasos, incluyendo factores como la aerolínea, el clima y otros.

En este DataFrame he reducido del conjunto de datos original a 200.000 para obtener resultados reproducibles. A partir de ahora se tomara en cuenta solo el DataFrame reducido para desarrollar los ejercicios.

In [11]:
#1.3: Columnas consideradas relevantes:

import pandas as pd 
df = pd.read_csv("C:/Users/acbon/Desktop/Data Science/DelayedFlights.csv")
random_seed = 42

columnas_relevantes = ['Origin', 'ArrDelay', 'CarrierDelay', 'WeatherDelay',
                       'NASDelay', 'SecurityDelay', 'LateAircraftDelay']

df_reducido_relevante = df_reducido[columnas_relevantes]
print(df_reducido_relevante.head())



        Origin  ArrDelay  CarrierDelay  WeatherDelay  NASDelay  SecurityDelay  \
1782417    IAD      71.0          71.0           0.0       0.0            0.0   
512712     ORD     548.0         543.0           0.0       0.0            0.0   
447137     SLC       9.0           NaN           NaN       NaN            NaN   
55082      HDN      52.0           0.0           0.0       0.0            0.0   
877634     JAX      22.0           0.0           0.0      15.0            7.0   

         LateAircraftDelay  
1782417                0.0  
512712                 5.0  
447137                 NaN  
55082                 52.0  
877634                 0.0  


Las columnas que he considerado relevantes son:


-Origin: El aeropuerto de origen.

-ArrDelay: El retraso en la llegada de los vuelos, ya que queremos saber cuánto se han retrasado.

-CarrierDelay: El retraso causado por la aerolínea.

-WeatherDelay: El retraso debido a condiciones climáticas.

-NASDelay: El retraso causado por el sistema de aviación nacional (NAS, por sus siglas en inglés).

-SecurityDelay: El retraso debido a problemas de seguridad.

-LateAircraftDelay: El retraso causado por un avión anterior que llegó tarde.

Estas columnas son esenciales para comprender y mejorar la puntualidad de los vuelos, la seguridad de la aviación y la eficiencia operativa en la industria de la aviación. También pueden ser de utilidad para los pasajeros y las autoridades reguladoras para garantizar una experiencia de vuelo más fluida y segura.





In [19]:
 #Ejercicio 2: Realiza un informe completo del dataset:

#Resume estadísticamente el dataset y las columnas de interés. Haz un análisis estadístico de lo que consideres relevante.
#Encuentra cuántos datos faltantes hay por columna.
#Crea nuevas columnas (velocidad media del vuelo, si ha llegado tarde o no...).
#Haz una mesa de las aerolíneas con mayores retrasos acumulados.
#¿Cuáles son los vuelos más largos? ¿Y los más atrasados? Busca las rutas más largas y las que acumulen más retrasos.
#Aporta lo que consideres relevante.

import pandas as pd

df = pd.read_csv("C:/Users/acbon/Desktop/Data Science/DelayedFlights.csv")
random_seed = 42
df_reducido = df.sample(n=200000, random_state=random_seed)

# 2.1: Resumen estadístico:
df_reducido.describe()



Unnamed: 0      Year          Month     DayofMonth      DayOfWeek  \
count  2.000000e+05  200000.0  200000.000000  200000.000000  200000.000000   
mean   3.338954e+06    2008.0       6.106975      15.734955       3.983700   
std    2.062108e+06       0.0       3.476865       8.770451       1.995799   
min    1.800000e+01    2008.0       1.000000       1.000000       1.000000   
25%    1.521904e+06    2008.0       3.000000       8.000000       2.000000   
50%    3.239627e+06    2008.0       6.000000      16.000000       4.000000   
75%    4.958252e+06    2008.0       9.000000      23.000000       6.000000   
max    7.009603e+06    2008.0      12.000000      31.000000       7.000000   

             DepTime     CRSDepTime        ArrTime     CRSArrTime  \
count  200000.000000  200000.000000  199267.000000  200000.000000   
mean     1520.114805    1468.567445    1611.293159    1635.514525   
std       450.087506     424.431628     548.219910     464.072463   
min         1.000000       0.000000       1.000000       0.000000   
25%      1205.000000    1135.000000    1317.000000    1325.000000   
50%      1546.000000    1510.000000    1716.000000    1707.000000   
75%      1901.000000    1817.000000    2031.000000    2015.000000   
max      2400.000000    2359.000000    2400.000000    2400.000000   

           FlightNum  ...      Distance         TaxiIn        TaxiOut  \
count  200000.000000  ...  200000.00000  199267.000000  199956.000000   
mean     2188.548195  ...     764.96174       6.817747      18.217083   
std      1944.372086  ...     572.85671       5.342789      14.410192   
min         1.000000  ...      11.00000       0.000000       0.000000   
25%       615.000000  ...     340.00000       4.000000      10.000000   
50%      1549.000000  ...     607.00000       6.000000      14.000000   
75%      3423.000000  ...     998.00000       8.000000      21.000000   
max      9740.000000  ...    4962.00000     200.000000     342.000000   

           Cancelled       Diverted   CarrierDelay   WeatherDelay  \
count  200000.000000  200000.000000  128970.000000  128970.000000   
mean        0.000300       0.004065      19.197666       3.691052   
std         0.017318       0.063628      44.126249      20.863373   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       2.000000       0.000000   
75%         0.000000       0.000000      21.000000       0.000000   
max         1.000000       1.000000    2436.000000     868.000000   

            NASDelay  SecurityDelay  LateAircraftDelay  
count  128970.000000  128970.000000      128970.000000  
mean       15.002566       0.091223          25.467403  
std        33.913939       2.014928          42.275946  
min         0.000000       0.000000           0.000000  
25%         0.000000       0.000000           0.000000  
50%         2.000000       0.000000           8.000000  
75%        15.000000       0.000000          34.000000  
max      1030.000000     280.000000        1012.000000  

[8 rows x 25 columns]


Entre los datos que muestra este dataset, he considerado relevante la siguiente información:

-Las estadísticas se basan en vuelos del año 2008.

-Los vuelos están distribuidos a lo largo de todo el año, los días del mes y los días de la semana.

-Los vuelos suelen despegar y aterrizar un poco más tarde de lo programado.

-La distancia promedio de vuelo es de aproximadamente 765 millas.

-Los tiempos promedio de taxi de entrada y salida son de aproximadamente 6.8 y 18.2 minutos, respectivamente.

-La mayoría de los vuelos no se cancelaron ni se desviaron.

-Los retrasos son raros, pero algunos vuelos experimentaron retrasos sustanciales debido a diversas causas, como el retraso de la aerolínea, retraso debido al clima, retraso del Sistema de Aviación Nacional, retraso de seguridad y retraso debido a un avión llegando tarde.


In [8]:
#2.2: Encuentra cuántos datos faltantes hay por columna.

import pandas as pd
import random  


df = pd.read_csv("C:/Users/acbon/Desktop/Data Science/DelayedFlights.csv")
random_seed = 42

datos_faltantes = df.isna().sum()


print("Datos faltantes por columna:")
print(datos_faltantes)


Datos faltantes por columna:
Unnamed: 0                0
Year                      0
Month                     0
DayofMonth                0
DayOfWeek                 0
DepTime                   0
CRSDepTime                0
ArrTime                7110
CRSArrTime                0
UniqueCarrier             0
FlightNum                 0
TailNum                   5
ActualElapsedTime      8387
CRSElapsedTime          198
AirTime                8387
ArrDelay               8387
DepDelay                  0
Origin                    0
Dest                      0
Distance                  0
TaxiIn                 7110
TaxiOut                 455
Cancelled                 0
CancellationCode          0
Diverted                  0
CarrierDelay         689270
WeatherDelay         689270
NASDelay             689270
SecurityDelay        689270
LateAircraftDelay    689270
dtype: int64


In [7]:
#2.3: Crea nuevas columnas (velocidad media del vuelo, si ha llegado tarde o no...)

import pandas as pd

df = pd.read_csv("C:/Users/acbon/Desktop/Data Science/DelayedFlights.csv")
df_reducido = df.sample(n=200000, random_state=random_seed)
random_seed = 42
df_reducido['VelocidadMedia'] = df_reducido['Distance'] / (df_reducido['AirTime']/60)
df_reducido['LlegadaTarde'] = df_reducido['ArrDelay'] > 15

print("Columna 'VelocidadMedia':")
print(df_reducido['VelocidadMedia'])

print("\nColumna 'LlegadaTarde':")
print(df_reducido['LlegadaTarde'])



Columna 'VelocidadMedia':
1782417    439.661017
512712     364.363636
447137     204.545455
55082      266.250000
877634     365.820896
              ...    
1106815    354.285714
269784     358.588235
627539     349.285714
292264     480.606061
554861     421.104294
Name: VelocidadMedia, Length: 200000, dtype: float64

Columna 'LlegadaTarde':
1782417     True
512712      True
447137     False
55082       True
877634      True
           ...  
1106815    False
269784      True
627539     False
292264     False
554861      True
Name: LlegadaTarde, Length: 200000, dtype: bool


In [6]:
#2.4: Haz una tabla de las aerolíneas con mayores retrasos acumulados.

import pandas as pd

df = pd.read_csv("C:/Users/acbon/Desktop/Data Science/DelayedFlights.csv")
df_reducido = df.sample(n=200000, random_state=random_seed)
random_seed = 42

retrasos_acumulados = df.groupby('UniqueCarrier')['ArrDelay'].sum()

aerolineas_con_retraso = retrasos_acumulados.sort_values(ascending=False)

tabla_retrasos_acumulados = pd.DataFrame({'Aerolínea': aerolineas_con_retraso.index, 'Retraso Acumulado': aerolineas_con_retraso.values})

print(tabla_retrasos_acumulados)

   Aerolínea  Retraso Acumulado
0         WN         11319092.0
1         AA          8889066.0
2         UA          6733013.0
3         MQ          6396704.0
4         OO          5978936.0
5         XE          5176042.0
6         DL          4535644.0
7         CO          4045932.0
8         EV          3888131.0
9         YV          3691461.0
10        US          3571867.0
11        NW          3462075.0
12        FL          3100150.0
13        B6          3025749.0
14        OH          2675993.0
15        9E          2420468.0
16        AS          1406735.0
17        F9           788549.0
18        HA           255613.0
19        AQ            15814.0


In [5]:
#2.5: ¿Cuáles son los vuelos más largos? ¿Y los más atrasados? Busca las rutas más largas y las que acumulen más retrasos.

import pandas as pd

df = pd.read_csv("C:/Users/acbon/Desktop/Data Science/DelayedFlights.csv")
df_reducido = df.sample(n=200000, random_state=random_seed)
random_seed = 42

rutas_mas_largas = df.groupby(['Origin', 'Dest'])['Distance'].sum()
rutas_mas_largas = rutas_mas_largas.sort_values(ascending=False)


retrasos_acumulados = df.groupby(['Origin', 'Dest'])['ArrDelay'].sum()
rutas_con_mas_retrasos = retrasos_acumulados.sort_values(ascending=False)

print("Rutas más largas:")
print(rutas_mas_largas.head(10))

print("\nRutas con más retrasos acumulados:")
print(rutas_con_mas_retrasos.head(10))


Rutas más largas:
Origin  Dest
LAX     JFK     5944950
SFO     JFK     5916768
JFK     LAX     5826150
ORD     LAX     5140770
LAX     ORD     4776065
JFK     SFO     4737552
ORD     SFO     4325178
SFO     ORD     4109196
ATL     LAX     4049626
LAX     HNL     3984804
Name: Distance, dtype: int64

Rutas con más retrasos acumulados:
Origin  Dest
ORD     LGA     241311.0
LAX     SFO     226548.0
ORD     EWR     216844.0
LGA     ORD     214032.0
ATL     EWR     191248.0
SFO     LAX     186803.0
ATL     LGA     186050.0
EWR     ORD     179692.0
DFW     ORD     169099.0
LGA     ATL     162820.0
Name: ArrDelay, dtype: float64


In [4]:

#2.6: Aporta lo que consideres relevante.


import pandas as pd

df = pd.read_csv("C:/Users/acbon/Desktop/Data Science/DelayedFlights.csv")
df_reducido = df.sample(n=200000, random_state=random_seed)
random_seed = 42

df_reducido['DepTime'] = df_reducido['DepTime'].fillna(0).astype(int)
df_reducido['ArrTime'] = df_reducido['ArrTime'].fillna(0).astype(int)
df_reducido['HoraSalida'] = df_reducido['DepTime'].apply(lambda x: f"{x//100:02d}:{x%100:02d}" if x > 0 else None)
df_reducido['HoraLlegada'] = df_reducido['ArrTime'].apply(lambda x: f"{x//100:02d}:{x%100:02d}" if x > 0 else None)

print(df_reducido[['HoraSalida', 'HoraLlegada']])







        HoraSalida HoraLlegada
1782417      18:05       21:46
512712       21:20       22:41
447137       17:31       18:19
55082        15:07       16:01
877634       06:37       08:12
...            ...         ...
1106815      19:03       19:57
269784       18:34       20:09
627539       16:33       17:50
292264       21:37       23:37
554861       18:48       21:48

[200000 rows x 2 columns]


He considerado importante también mostrar las horas de salida y llegada en formato legible. Esta información puede ser utilizada para varios propósitos, dependiendo de las necesidades de análisis o investigación. Por ejemplo, se pueden crear visualizaciones de patrones horarios mediante gráficos o representaciones visuales que muestren cuándo tienden los vuelos a estar más retrasados o más puntuales durante el día.



In [10]:
#Ejercicio 3:Exporta el dataset limpio y con las nuevas columnas a Excel.

import pandas as pd 
df = pd.read_csv("C:/Users/acbon/Desktop/Data Science/DelayedFlights.csv")
random_seed = 42

ruta_salida = "C:/Users/acbon/Desktop/Data Science/DelayedFlights.xlsx"

df_reducido.to_excel(ruta_salida, index=False)

print(f"El archivo Excel se ha exportado a: {ruta_salida}")


El archivo Excel se ha exportado a: C:/Users/acbon/Desktop/Data Science/DelayedFlights.xlsx
